### PR TITLE
Encoding/decoding of date/time/datetime

### PR DESCRIPTION
**Pull request by jquadrin**
continues #18

### CODE REVIEW

**Comment by laurocaetano** [comment]
@jquadrin nice work and thanks for taking a look into these types :heart: 

I'll close my PR :)

**Comment by josevalim** [comment]
Those types unfortunately are not going to work with Ecto. Ecto expects those types to be tuples, so we should revert to the previous definitions. There is currently a discussion about moving from tuples to maps, but I believe we should get the adapter up to date first and the discuss those changes, as it will also involved TDS and Postgrex adapters.

File: lib/mariaex/messages.ex [hunk 1]
```diff
@@ -264,13 +233,103 @@ defmodule Mariaex.Messages do
...
+  defp decode_date(data) do 
+    [year, month, day] = '~d-~d-~d'
+    |> :io_lib.fread(to_char_list(data)) 
+    |> elem(1)
+    %Mariaex.Date{year: year, month: month, day: day}
+  end
+
+  defp decode_time(data) do
+    [hour, min, sec] = '~d:~d:~d'
+    |> :io_lib.fread(to_char_list(data))
```
Comment (josevalim) [inline lib/mariaex/messages.ex]: I think `String.split(data, ":") |> Enum.map(&String.to_integer/1)` will be cleaner and faster!

File: lib/mariaex/messages.ex [hunk 2]
```diff
@@ -264,13 +233,103 @@ defmodule Mariaex.Messages do
...
+  defp decode_msg(body, :rows),                                                    do: __decode__(:row, body)
+
+  defp decode_string(data),    do: data
+  defp decode_float(data),     do: Float.parse(data) |> elem(0)
+  defp decode_integer(data),   do: Integer.parse(data) |> elem(0)
+  defp decode_bit(<<bit>>),    do: bit
+  defp decode_null(_),         do: nil
+
+  defp decode_date(data) do 
+    [year, month, day] = '~d-~d-~d'
```
Comment (josevalim) [inline lib/mariaex/messages.ex]: I think String.split(data, "-") |> Enum.map(&String.to_integer/1) will be cleaner and faster!

**Comment by josevalim** [comment]
This is a nice cleaned! I have added some suggestions too. We should definitely rollback to the previous types though, at least for now.

**Comment by jquadrin** [comment]
thanks José! :smiley:

**Comment by josevalim** [comment]
Fantastic work both! This looks good to be merged from my side too!

**Comment by liveforeverx** [comment]
Huge thanks! This looks very good!

**Comment by liveforeverx** [comment]
Sorry for delay, have at the moment(hopefully for max next 3-4 days) much work.

**Pull request merged**